design-proposals/klip-50-partition-and-offset-in-ksqldb.md

```diff
@@ -0,0 +1,253 @@
+			("poison pills") during processing._
+
+## Motivation and background
+
+As described in the relevant [Github issue](https://github.com/confluentinc/ksql/issues/976),
+it's helpful when developing and troubleshooting ksqlDB queries, or when using ksqlDB to
+inspect other anomalies in a Kafka topic, to be able to target individual Kafka records.
+Today this can be approximated by filtering on the message timestamp via the `ROWTIME`
+pseudocolumn in ksqlDB, but this is only approximate as there could be multiple messages
+with the same timestamp. In contrast, the combination of topic partition and offset is
+guaranteed to uniquely identify a specific message.
+
+## What is in scope
+
+* Syntax for accessing Kafka topic partition and offset
+* Desired query semantics (persistent and transient)
+* Desired pull query semantics (specifically, that these metadata will not be available 
+  by default and must instead be copied into the message value to be accessible)
+* Other semantics, such as whether these metadata will be included in source descriptions
+  or `PRINT TOPIC` commands, or allowed in `INSERT VALUES` statements
+* Compatibility implications for existing ksqlDB queries
+* High-level implementation strategy
+
+## What is not in scope
+
+* [Exposing Kafka message headers in ksqlDB](https://github.com/confluentinc/ksql/issues/1940).
+  We would like to do this in the future but we'll have a separate KLIP for this as it's a
+  larger discussion.
+
+## Value/Return
+
+ksqlDB will expose message partition and offset as pseudocolumns similar to the `ROWTIME`
+pseudocolumn that exists today. Users will be able to select the partition and offset in
+transient queries, persistent queries, and also use the metadata in filters and UDFs.
+(See ["Public APIS"](#public-apis) below for example syntax.)
+
+This will allow users to more easily inspect data and debug ksqlDB queries:
+```
+SELECT *, ROWPARTITION, ROWOFFSET FROM my_stream EMIT CHANGES;
+```
+experiment with processing only a subset of messages:
+```
+SELECT * FROM my_stream WHERE ROWPARTITION = 0 AND ROWOFFSET > 12 EMIT CHANGES;
+```
+and skip select messages ("poison pills") during processing:
+```
+CREATE STREAM my_cleaned_stream AS SELECT * FROM my_stream WHERE NOT (ROWPARTITION = 0 AND ROWOFFSET = 42);
+```
+
+## Public APIS
+
+We will add two new [pseudocolumns](https://github.com/confluentinc/ksql/blob/master/design-proposals/klip-14-rowtime-as-pseudocolumn.md)
+to ksqlDB: `ROWPARTITION` (integer) and `ROWOFFSET` (bigint). We choose these names to be 
+consistent with the existing `ROWTIME` pseudocolumn, and to minimize the chance of conflict 
+with users' existing data columns for compatibility reasons. See 
+["Alternatives Considered"](#alternatives-considered) for discussion on possible alternatives.
+
+### Query Semantics
+
+The behavior of `ROWPARTITION` and `ROWOFFSET` in persistent queries and push queries
```

Comment:
From the section below, seems you were proposing that we do not add these two to the push queries, right? If yes maybe better to remove `and push queries` here.